---
title: "Overlap with chip-seq"
author: "wancen"
date: "2/6/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
## Load required libraries
suppressMessages(suppressWarnings(library(data.table)))
suppressMessages(suppressWarnings(library(plyranges)))
suppressMessages(suppressWarnings(library(tidyverse)))
suppressMessages(suppressWarnings(library(devtools)))
load_all("..")
```


```{r}
mESC<-read_narrowpeaks("../data/mESC.narrowPeak",genome_info = "mm9")
colnames(mcols(mESC))<-c("name","score","fold_enrichment","-log(pvalue)","-log(FDR)/log(10)","summit")
# mESC<-mESC[which(mcols(mESC)$fold_enrichment>20)]

mef<-read_narrowpeaks("../data/mef.narrowPeak",genome_info = "mm9")
colnames(mcols(mef))<-c("name","score","fold_enrichment","-log(pvalue)","-log(FDR)/log(10)","summit")
# mef<-mef[which(mcols(mef)$fold_enrichment>20)]

suppressMessages(suppressWarnings(library(GenomicFeatures)))
suppressMessages(suppressWarnings(library("org.Mm.eg.db")))
txdb <- makeTxDbFromGFF("ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M1/gencode.vM1.annotation.gtf.gz")
g <- genes(txdb)
g$sym <- mapIds(org.Mm.eg.db, sub("\\..*","",names(g)), "SYMBOL", "ENSEMBL")
g <- g[!is.na(g$sym)]
# define promoter regions
gene_prom <- promoters(x = g, upstream = 20000, downstream = 300)


## Get allelic imbalance genes
get_overlap<-function(gene_prom,feat,mef,mESC){
  poi<-which(mcols(gene_prom)$sym %in% names(feat))
  g2<-gene_prom[poi]
  
  
  mef_ov<-join_overlap_inner(g2, mef)
  mESC_ov<-join_overlap_inner(g2, mESC)
  
  mef_ov2 <- mef_ov %>%
    group_by(sym) %>%
    summarize(
      fold_enrichment=sum(fold_enrichment),
      peak_count=sum(!is.na(fold_enrichment)))
  
  mESC_ov2 <- mESC_ov %>%
    group_by(sym) %>%
    summarize(
      fold_enrichment=sum(fold_enrichment),
      peak_count=sum(!is.na(fold_enrichment)))
  
  enrich<-merge(mESC_ov2,mef_ov2,by="sym",all=T)
  return(enrich)
}
```

```{r}
load("../data/larsson_postprocess.rda")
nct<-length(cell_meta_unique) # number of cell types
head(total_fil[1:5,1:5])
ratio<-c57_fil/total_fil
ratio_pseudo<-(c57_fil+2)/(total_fil+4)
cluster<-genecluster(ratio_pseudo,nct,G=c(4,12,20,24,28,32))
table(cluster)
```

```{r}
for (i in 1:28) {

  feat<-which(cluster==i) # choose genes location
  cl_ratio<-as.vector(unlist(ratio[feat,]))
  cl_total<-as.vector(unlist(total_fil[feat,]))
  dat<-tibble(ratio=cl_ratio,x=factor(rep(as.vector(celltypeQC[,1]),each=length(feat)),levels = cell_meta_unique),cts=cl_total)
  summary<-dat %>% group_by(x) %>% summarise(weighted.mean=weighted.mean(ratio,cts,na.rm = T),mean=mean(ratio,na.rm = T),var=var(ratio,na.rm = T)) %>% data.frame()
  summary2<-summary[order(summary$weighted.mean),]
  if(max(diff(summary2$weighted.mean))>0.2){
    print(paste("cluster",i,"has",length(feat),"genes exists allelic ratio imbalance"))
    
    print(knitr::kable(summary)) #check summary table
    overlaptxb<-get_overlap(gene_prom,feat,mef,mESC)
    print(knitr::kable(overlaptxb))
    
  }  

}
```

